File: backend/community_diagnostics.py
"""Structural diagnostics for the Community folder - catches two common,
hard-to-self-diagnose install mistakes that scenery_data.py's own scan
silently ignores (by design - it only cares about recognized packages):

1. Misplaced manifest: a folder with no manifest.json directly inside it,
   but one exactly one level deeper (Community/name/name/manifest.json
   instead of Community/name/manifest.json) - MSFS itself ignores these
   too, and it is a very common mistake when manually extracting a zip
   that already contains a wrapping folder.
2. Duplicate installs: the same package (by manifest title) present under
   two or more different folder names, anywhere across Community and the
   known disabled-holding locations - wastes disk space and can conflict.

Pure read-only diagnostics - never moves, renames, or deletes anything.
Kept separate from scenery_data.py (untouched) since this is a different
concern (structural sanity, not content_type-aware enable/disable) and not
limited to SCENERY packages the way that module is used elsewhere.
"""
import json
import os


def _read_manifest_title(folder_path):
    manifest_path = os.path.join(folder_path, "manifest.json")
    if not os.path.exists(manifest_path):
        return None
    try:
        with open(manifest_path, "r", encoding="utf-8-sig") as f:
            data = json.load(f)
    except Exception:
        return None
    return (data.get("title") or "").strip() or None


def find_misplaced_packages(community_path):
    """Returns [{"folder_name": str, "nested_folder": str}, ...] for
    top-level Community folders with no manifest.json of their own, but a
    valid one exactly one level deeper."""
    misplaced = []
    try:
        entries = list(os.scandir(community_path))
    except OSError:
        return misplaced

    for entry in entries:
        if not entry.is_dir():
            continue
        if os.path.exists(os.path.join(entry.path, "manifest.json")):
            continue  # a normal, correctly-placed package - nothing wrong here

        try:
            subentries = list(os.scandir(entry.path))
        except OSError:
            continue
        for sub in subentries:
            if sub.is_dir() and os.path.exists(os.path.join(sub.path, "manifest.json")):
                misplaced.append({"folder_name": entry.name, "nested_folder": sub.name})
                break

    return misplaced


def find_duplicate_packages(locations):
    """locations: folder paths to scan (Community + all known disabled-
    holding locations). Flags any manifest title found under 2+ distinct
    folder names anywhere across them. Returns [{"title": str,
    "folders": [folder_name, ...]}, ...]."""
    by_title = {}
    for base_path in locations:
        try:
            entries = list(os.scandir(base_path))
        except OSError:
            continue
        for entry in entries:
            if not entry.is_dir():
                continue
            title = _read_manifest_title(entry.path)
            if title:
                by_title.setdefault(title, set()).add(entry.name)

    return [
        {"title": title, "folders": sorted(folders)}
        for title, folders in by_title.items()
        if len(folders) > 1
    ]

File: dotnet/src/FlightOpsHub.Core/CommunityDiagnostics.cs
using System.Text.Json.Nodes;

namespace FlightOpsHub.Core;

/// <summary>
/// Port of backend/community_diagnostics.py - structural diagnostics for
/// the Community folder, catching two common install mistakes that
/// SceneryScanner silently ignores by design (it only cares about
/// recognized packages):
///
/// 1. Misplaced manifest: a folder with no manifest.json directly inside
///    it, but one exactly one level deeper (a very common mistake when
///    manually extracting a zip that already contains a wrapping folder).
/// 2. Duplicate installs: the same package (by manifest title) present
///    under two or more different folder names, anywhere across Community
///    and the known disabled-holding locations.
///
/// Pure read-only diagnostics - never moves, renames, or deletes anything.
/// </summary>
public static class CommunityDiagnostics
{
    private static string? ReadManifestTitle(string folderPath)
    {
        var manifest = ManifestReader.Read(folderPath);
        var title = ManifestReader.GetString(manifest, "title").Trim();
        return title.Length > 0 ? title : null;
    }

    /// <summary>Top-level Community folders with no manifest.json of their own, but a valid one exactly one level deeper.</summary>
    public static List<JsonObject> FindMisplacedPackages(string communityPath)
    {
        var misplaced = new List<JsonObject>();
        IEnumerable<string> entries;
        try
        {
            entries = Directory.EnumerateDirectories(communityPath).ToList();
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            return misplaced;
        }

        foreach (var entryPath in entries)
        {
            if (File.Exists(Path.Combine(entryPath, "manifest.json")))
            {
                continue; // a normal, correctly-placed package - nothing wrong here
            }

            IEnumerable<string> subentries;
            try
            {
                subentries = Directory.EnumerateDirectories(entryPath);
            }
            catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
            {
                continue;
            }

            foreach (var subPath in subentries)
            {
                if (File.Exists(Path.Combine(subPath, "manifest.json")))
                {
                    misplaced.Add(new JsonObject
                    {
                        ["folder_name"] = Path.GetFileName(entryPath),
                        ["nested_folder"] = Path.GetFileName(subPath),
                    });
                    break;
                }
            }
        }
        return misplaced;
    }

    /// <summary>
    /// locations: folder paths to scan (Community + all known disabled-
    /// holding locations). Flags any manifest title found under 2+
    /// distinct folder names anywhere across them.
    /// </summary>
    public static List<JsonObject> FindDuplicatePackages(IEnumerable<string> locations)
    {
        var byTitle = new Dictionary<string, SortedSet<string>>();
        foreach (var basePath in locations)
        {
            IEnumerable<string> entries;
            try
            {
                entries = Directory.EnumerateDirectories(basePath).ToList();
            }
            catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
            {
                continue;
            }

            foreach (var entryPath in entries)
            {
                var title = ReadManifestTitle(entryPath);
                if (title is null) continue;
                if (!byTitle.TryGetValue(title, out var folders))
                {
                    folders = new SortedSet<string>(StringComparer.Ordinal);
                    byTitle[title] = folders;
                }
                folders.Add(Path.GetFileName(entryPath));
            }
        }

        return byTitle
            .Where(kv => kv.Value.Count > 1)
            .Select(kv => new JsonObject
            {
                ["title"] = kv.Key,
                ["folders"] = new JsonArray(kv.Value.Select(f => (JsonNode)f).ToArray()),
            })
            .ToList();
    }
}
